### DISCUSSION TITLE
Questions about using advisory locks with prisma (postgres)

### DISCUSSION THREAD

**Discussion by dyuan-boxlabs** [prisma/prisma#16740] [Q&A]
Question 1:
Since prisma handles connection pooling for you, and because postgresql's advisory locks are session based, it prevents you from using the session based advisory locks because you can't guarantee the unlock happens on the same connection/session correct?

Example:
`await prisma.$queryRawUnsafe(`select pg_advisory_lock(1)`);`
`// do stuff`
`// might not be on the same connection, so you end up with a dead lock until the original session terminates`
`await prisma.$queryRawUnsafe(`select pg_advisory_unlock(1)`);`

Question 2:
The only way to do this is within a prisma transaction - in which case is a transaction advisory lock more appropriate (auto unlock at commit)?

Example:
`await prisma.$transaction(async tx => {`
  `await tx.$queryRawUnsafe(`select pg_advisory_xact_lock(1)`);`
  `// do stuff`
 `// auto unlock`
`})`

**Comment by nurul3101** [comment]
Hey @dyuan-boxlabs 👋 

You can use session-based advisory locks if you set [connection_limit](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/connection-pool#setting-the-connection-pool-size) to 1 in your connection string, it might not be the best approach, but you could allow only one database connection to PrismaClient and that would ensure that the same connection gets used.

However, using transactions would be an ideal choice, you can start lock at transaction beginning and release it at the end of transaction.

**Reply by hco** [reply]
> However, using transactions would be an ideal choice, you can start lock at transaction beginning and release it at the end of transaction.

This is kind of in conflict with the recommendation on https://www.prisma.io/docs/orm/prisma-client/queries/transactions#interactive-transactions, which says:

> Use interactive transactions with caution. Keeping transactions open for a long time hurts database performance and can even cause deadlocks. Try to avoid performing network requests and executing slow queries inside your transaction functions. We recommend you get in and out as quick as possible!

Do you agree? I am still struggeling with finding a good way of using advisory locks together with prisma :-/

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
